gmt: 6.4.0 -> 6.5.0

doronbehar opened the pull request:
## Description of changes


## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar commented:
@tviti please observe the link phase error:

```
/nix/store/d828ccvc2148g7m49hh3mzvyzwpipy46-binutils-2.42/bin/ld: /nix/store/wmlahsm7jjkxi69p0nhil35l63964acn-libspatialite-5.1.0/lib/libspatialite.so.8: undefined reference to `xmlNanoHTTPCleanup@LIBXML2_2.4.30'
```

sikmir commented:
> Maybe someone from the @NixOS/geospatial members can help here? I'm trying to fix the link phase error in `gmt`, which uses `libspatialite`, the error is:
> 
> ```
> /nix/store/d828ccvc2148g7m49hh3mzvyzwpipy46-binutils-2.42/bin/ld: /nix/store/wmlahsm7jjkxi69p0nhil35l63964acn-libspatialite-5.1.0/lib/libspatialite.so.8: undefined reference to `xmlNanoHTTPCleanup@LIBXML2_2.4.30'
> ```

I guess `(libxml2.override { enableHttp = true; })` somewhere required.

sikmir commented on pkgs/applications/gis/gmt/default.nix:
```diff
@@ -1,59 +1,98 @@
+    owner = "GenericMappingTools";
+    repo = "gmt";
+    rev = "refs/tags/${finalAttrs.version}";
+    sha256 = "sha256-KKIYhljCtk9t9CuvTLsSGvUkUwazWTm9ymBB3wLwSoI=";
```
```suggestion
    hash = "sha256-KKIYhljCtk9t9CuvTLsSGvUkUwazWTm9ymBB3wLwSoI=";
```

doronbehar commented:
Thanks for trying to help @sikmir . Unfortunately, it didn't help :/. Will try to ask for help from @K900 per https://github.com/NixOS/nixpkgs/pull/330337 .

K900 commented:
Something else could also be propagating an unpatched libxml? But honestly this is bad, because libxml upstream is dropping HTTP support soon, so all of this will just break again.

doronbehar commented:
Are you sure that propagating a normal `libxml2` is the only thing that can cause this failure? Because `netcdf` uses the normal `libxml2` but it doesn't propagate it. I tried to override that too but it didn't help:

```nix
(netcdf.override { libxml2 = libxml2.override { enableHttp = true; }; })
```

sikmir commented:
https://github.com/NixOS/nixpkgs/pull/344941

autra committed:
python312Packages.pygmt: unbreak on Darwin

autra commented:
While we're at it, can you add:
```
  passthru.updateScript = gitUpdater {};
```
somewhere, so that the update bot gives us updates?

doronbehar committed:
gmt: modernize

Nixfmt; Use `finalAttrs:` pattern; don't use `with lib;` in meta; Switch
to fetchFromGitHub, for better automatic updates, with hash instead of
sha256.

doronbehar committed:
gmt: fix build by overriding netcdf

doronbehar committed:
gmt: 6.4.0 -> 6.5.0

Diff: https://github.com/GenericMappingTools/gmt/compare/refs/tags/6.4.0...6.5.0

doronbehar committed:
gmt: let it find libraries automatically

doronbehar commented:
Thanks @autra for linking your PR. Indeed what I missed with the `netcdf.override` was the usage of the unoverriden netcdf in the `cmakeFlags`. I incorporated your changes, but slightly differently, along with other improvements.

As for using `passthru.updateScript`, this won't be needed once we use `fetchFromGitHub`, as all automatic updating programs are capable of detecting new GitHub releases reliably.

doronbehar commented:
Unfortunately, the original motivation I had for fixing gmt was trying out `gmt histogram`:

https://docs.generic-mapping-tools.org/latest/histogram.html

It doesn't work:

```
$ ./result/bin/gmt histogram table.txt
gmt [ERROR]: Shared GMT module not found: histogram

ERROR: No module named histogram was found.  This could mean one of four cases:
  1. There actually is no such module; please check your spelling.
  2. You used a modern mode module name while running in GMT classic mode.
  3. Module exists in the GMT supplemental library, but the library could not be found.
  4. Module exists in a GMT custom library, but none was specified via GMT_CUSTOM_LIBS.
Shared libraries must be in standard system paths or set via environmental parameter LD_LIBRARY_PATH.
```

But if I run:

```
$ ./result/bin/gmt histogram
gmt histogram [core] 6.5.0 [64-bit] [MP] - Calculate and plot histograms

usage: gmt histogram [<table>] -Jx|X<args> -T[<min>/<max>/]<inc>[+i|n] [-A] [-B<args>] [-C<cpt>[+b]]
...
```

@autra do you use gmt directly? or only as a dependency? Maybe you'd have an idea how to install this histogram module? It might be a cmake flag missing ...

autra commented:
> As for using passthru.updateScript, this won't be needed once we use fetchFromGitHub, as all automatic updating programs are capable of detecting new GitHub releases reliably.

are you sure about that? I have some counter-examples, for example Ifcopenshell. Maybe there are some conditions (like the format of the tag or something)?

> @autra do you use gmt directly? or only as a dependency? Maybe you'd have an idea how to install this histogram module? It might be a cmake flag missing ...

Ah no, it's just a broken package that popped up when trying to update geos (which one on the dependencies of gmt), sorry!

autra commented on pkgs/applications/gis/gmt/default.nix:
```diff
@@ -1,59 +1,98 @@
+let
+  # Certainly not an ideal situation, See:
+  # https://github.com/NixOS/nixpkgs/pull/340707#issuecomment-2361894717
+  netcdf' = netcdf.override {
```
I think you don't need that now, because you've already overriden it in all-packages.

doronbehar replied:
> I think you don't need that now, because you've already overriden it in all-packages.

Thanks for catching that. I meant to only override things inside the expression.

doronbehar commented:
> for example Ifcopenshell.

Yes this is a particular example where unfortunately the automatic updaters fail to detect the version scheme there:

https://github.com/NixOS/nixpkgs/blob/7b067a12a807aa58ac65d397f15db7cd4af8316e/pkgs/development/python-modules/ifcopenshell/default.nix#L21-L27

See also: https://github.com/Mic92/nix-update/issues/9#issuecomment-1058215873 . For such packages, it seems currently unavoidable to explicitly specify an automatic updating method.

> Ah no, it's just a broken package that popped up when trying to update geos (which one on the dependencies of gmt), sorry!

OK I see. Do you approve the changes? I don't expect the maintainers to respond, as they don't seem to care that their package was broken for more then a month.

doronbehar commented:
With your permission, I also dropped a commit changing the maintainers of pygmt. I don't feel comfortable to commit that.

autra commented:
For the automatic update, I still have doubts 😅 . Without the `updateScript`, this command fails:

```
nix-shell maintainers/scripts/update.nix --argstr package gmt
```

Are you sure it works for non-python packages?  Because I know python packages get an automatic `passthru.updateScript` somewhere (I could find the exact line), but AFAIK not the others. Maybe I'm misunderstanding how ofborg detects these updates though. Do you know how it does?

> OK I see. Do you approve the changes? I don't expect the maintainers to respond, as they don't seem to care that their package was broken for more then a month.

It seems fine, but @sikmir is the current maintainer and has already participated in this thread 😆 So I let him approve... 

For the maintainer change, we can see later if @sikmir approves to have this under the whole geospatial team (which he is part of ).

doronbehar commented:
> For the automatic update, I still have doubts 😅 . Without the `updateScript`, this command fails:

Haha I'm sorry, but when I said "automatic updaters", I meant only [`nix-update`](https://github.com/Mic92/nix-update) and [`nixpkgs-update`](https://github.com/nix-community/nixpkgs-update). `nix-shell maintainers/scripts/update.nix --argstr package gmt` is an update script, but I wouldn't call it automatic one :).

> Maybe I'm misunderstanding how ofborg detects these updates though. Do you know how it does?

Are you wondering how ofborg detects which packages to build when it runs it's evaluation on a PR? Because that is [documented in it's README](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building), but ofborg doesn't commit any updates at all.

> It seems fine, but @sikmir is the current maintainer and has already participated in this thread 😆 So I let him approve...
> 
> For the maintainer change, we can see later if @sikmir approves to have this under the whole geospatial team (which he is part of ).

OK, you can still give your approval, and we'll give him until next week to respond. I suspect that he has done the minimum required, which is not obvious at all, which is to notify me that another gmt related PR (yours) is happening, but I still suspect he cannot spare time to review this. Let's hope that I'm wrong :).

sikmir commented:
Please reword commit message `pygmt: ...` to `python312Packages.pygmt: ...`.

doronbehar commented:
> Please reword commit message `pygmt: ...` to `python312Packages.pygmt: ...`.

Thanks :). Done.

sikmir commented:
Result of `nixpkgs-review pr 340707` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>gmt</li>
    <li>python311Packages.pygmt</li>
    <li>python311Packages.pygmt.dist</li>
    <li>python312Packages.pygmt</li>
    <li>python312Packages.pygmt.dist</li>
  </ul>
</details>

sikmir commented:
Result of `nixpkgs-review pr 340707` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>gmt</li>
    <li>python311Packages.pygmt</li>
    <li>python311Packages.pygmt.dist</li>
    <li>python312Packages.pygmt</li>
    <li>python312Packages.pygmt.dist</li>
  </ul>
</details>

doronbehar commented:
Thanks @sikmir for the review :).
